qc-runner/src/cellSizeDistribution.r

```diff
@@ -74,29 +74,30 @@ task <- function(seurat_obj, config, task_name, sample_id, num_cells_to_downsamp
         return(list(data = seurat_obj,config = config,plotData = plots)) 
     }
     sample_subset <- subset(seurat_obj, cells = barcode_names_this_sample)
-    plot1_data  <- sort(sample_subset$nCount_RNA, decreasing = TRUE)
-    # plot 1: histgram of UMIs, hence input is all UMI values, e.g.
-    # AAACCCAAGCGCCCAT-1 AAACCCAAGGTTCCGC-1 AAACCCACAGAGTTGG-1
-    #               2204              20090               5884  ..   
-    #    u    u    u    u    u    u
-    # 3483 6019 3892 3729 4734 3244
-    
-    plot2_data <- seq_along(plot1_data)
-    plot1_data_transformed <- log(plot1_data)
-    plot2_data <- unname(map2(plot1_data_transformed,plot2_data,function(x,y){c("log_u"=x,"rank"=y)}))
-    plot1_data <- lapply(unname(plot1_data),function(x) {c("u"=x)})
-
-    # Downsample plotData
-    # Handle when the number of remaining cells is less than the number of cells to downsample
-    num_cells_to_downsample <- downsample_plotdata(ncol(sample_subset), percent_downsample, num_cells_to_downsample)
-    print(paste('sample of size', ncol(sample_subset), 'downsampled to', num_cells_to_downsample, 'cells'))
-    
+
+    # umi histogram plot
+    numis <- sort(sample_subset$nCount_RNA, decreasing = TRUE)
+    plot1_data <- lapply(unname(numis),function(x) {c("u"=x)})
+
+    # barcode ranks plot data
+    # unique average ranks maintain plot shape in case of downsampling
+    ranks <- rank(-numis, ties.method = 'average')
+    dups <- duplicated(ranks)
+    ranks <- ranks[!dups]
```

Comment:
Whats the rational to remove duplicated ranks? Is it for cosmetic reasons? I'm just concerend that this would bias the plot, i.e. in low abundant cells all with say a count of 11 would be condensed to only 1 cell.

Reply:
The rationale is that e.g. the we could have ~600k drops with the majority being <10 UMIs. These droplets have the same rank and the same UMI so they have the exact same coordinates on the kneeplot. By removing them we can get down to ~8k cells and then then subsequent downsampling to 6k will still yield a kneeplot with the exact shape instead of being heaving skewed towards high-rank low UMI droplets (and dropping the high UMI low rank cells we are interested in keeping so that they contribute to the shape). We only do this for the knee plot.

Reply:
Are we erasing the repeated ranks after or before we calculate the inflection points? This should be fine if we are doing it post calculation of the inflection points. Also, we need to make sure that this change to the ranks doesn't negatively impact the positioning of the vertical threshold line in the plot bc it's inferred from the min cell size.

Reply:
We are erasing the repeated rank only for the knee plot data. The calculation of the inflection point is unchanged

Reply:
Great! Thanks for clarifying alex, I'll make sure we aren't unintentionally breaking the UI threshold line and take a look at the staging links and if all looks good I'll approve and merge.